packages/vite/src/node/plugins/worker.ts

```diff
@@ -42,61 +71,100 @@ export function webWorkerPlugin(config: ResolvedConfig): Plugin {
+      const workerConstructor = buildWorkerConstructor(query)
+      if (query == null || workerConstructor == null) {
         return
       }
 
       let url: string
       if (isBuild) {
-        if (query.inline != null) {
-          // bundle the file as entry to support imports and inline as blob
-          // data url
-          const rollup = require('rollup') as typeof Rollup
-          const bundle = await rollup.rollup({
-            input: cleanUrl(id),
-            plugins: config.plugins as Plugin[]
+        // bundle an inline script version of the worker
+        const { rollup } = require('rollup') as typeof Rollup
+        const bundle = await rollup({
+          input: cleanUrl(id),
+          plugins: config.plugins as Plugin[]
+        })
+
+        let code: string
+        try {
+          const { output } = await bundle.generate({
+            format: 'es',
+            sourcemap: config.build.sourcemap
           })
-          try {
-            const { output } = await bundle.generate({
-              format: 'es',
-              sourcemap: config.build.sourcemap
-            })
-            
-            return `const blob = new Blob([atob(\"${Buffer.from(output[0].code).toString('base64')}\")], { type: 'text/javascript;charset=utf-8' });
-            export default function WorkerWrapper() {
-              const objURL = (window.URL || window.webkitURL).createObjectURL(blob);
-              try {
-                return new Worker(objURL);
-              } finally {
-                (window.URL || window.webkitURL).revokeObjectURL(objURL);
-              }
-            }`
-          } finally {
-            await bundle.close()
-          }
+
+          code = output[0].code
+        } finally {
+          await bundle.close()
+        }
+        const content = Buffer.from(code)
+
+        if (query.inline != null) {
+          // inline as blob data url
+          return `const blob = new Blob([atob(\"${content.toString(
+            'base64'
+          )}\")], { type: 'text/javascript;charset=utf-8' });
+          const URL = window.URL || window.webkitURL;
+          export default function WorkerWrapper() {
+            const objURL = URL.createObjectURL(blob);
+            try {
+              return ${workerConstructor('objUrl')};
+            } finally {
+              URL.revokeObjectURL(objURL);
+            }
+          }`
         } else {
-          // emit as separate chunk
+          // emit as separate chunk (type module)
           url = `__VITE_ASSET__${this.emitFile({
             type: 'chunk',
             id: cleanUrl(id)
           })}__`
+
+          // emit separate bundled chunk (type script)
+          const basename = path.basename(cleanUrl(id))
+          const ext = path.extname(basename)
+          const contentHash = getAssetHash(content)
+          const fileName = path.posix.join(
+            config.build.assetsDir,
+            `${basename.slice(0, -ext.length)}.${contentHash}.js`
```

Comment:
`basename.slice(0, -ext.length)` can be simplified to `path.parse(cleanUrl(id)).name`